data-reconciliation-app/docs/adr/10-swagger-endpoint.md

```diff
@@ -0,0 +1,154 @@
+
+```
+Content-Security-Policy: script-src 'sha256-abc123' 'sha512-321cba'
+```
+
+In the presence of that policy, the following script elements would be allowed to execute because they contain only integrity metadata that matches the policy:
+
+```
+<script integrity="sha256-abc123" ...></script>
+<script integrity="sha512-321cba" ...></script>
+<script integrity="sha256-abc123 sha512-321cba" ...></script>
+```
+
+While the following script elements would not execute because they contain valid metadata that does not match the policy (even though other metadata does match):
+
+```
+<script integrity="sha384-xyz789" ...></script>
+<script integrity="sha384-xyz789 sha512-321cba" ...></script>
+<script integrity="sha256-abc123 sha384-xyz789 sha512-321cba" ...></script>
+```
+
+To implement this option, we have to calculate the `hash` of all external file content and include in the headers:
+
+```bash
+echo -n 'doSomething();' | openssl sha256 -binary | openssl base64
+```
+
+The `hash` will be then attached to the script tags respectively.
+
+```html
+<script integrity="sha256-abc123" ...></script>
+```
+----
+
+### 2. Import npm package and serve from TS
+
+> **⚠️ This option requires further investigation on server side rendering of this package. As we are building an API and not a SPA, due to Swagger UI doesn't support server side rendering functionality, we will disregard this option.**
+
+[Swagger UI](https://www.npmjs.com/package/swagger-ui) documentation strongly suggest that we use `swagger-ui` instead of `swagger-ui-dist` since swagger-ui-dist is significantly larger. They suggest using `swagger-ui` when the tooling makes it possible, as `swagger-ui-dist` will result in more code going across the wire.
+
+Using [Swagger UI](https://www.npmjs.com/package/swagger-ui), we can implement this package in our endpoint if we were to build a SPA.
+
+```typescript
+import * as ccfapp from "@microsoft/ccf-app";
+import { SwaggerUIBundle } from 'swagger-ui-dist';
+
+/**
+ * Generate a swagger UI based on the OpenApi documents of (Application - Governance)
+ * How to configure swagger UI: https://swagger.io/docs/open-source-tools/swagger-ui/usage/installation/
+ */
+export function getSwaggerUI(): ccfapp.Response<string> {
+
+  const ui = SwaggerUIBundle({
+      urls: [
+        {url: "/app/swagger.json", name: "Application API"}, /* '/app/api' will be used when ccf allow more control on openAPI document to be done through deployment bundle */
+        {url: "/gov/api", name: "Governance API"},
+        {url: "/node/api", name: "Operator API"}
+      ],
+      dom_id: '#swagger-ui',
+      presets: [SwaggerUIBundle.presets.apis],
+      layout: "StandaloneLayout",
+    });
+
+    return ui;
+};
+```
+
+### 3. Download external script and serve from the local repo
+
+This option is implemented by simply downloading the following script and css files and commit to this repo. These files will be imported to the HTML by a relative path.
```

Comment:
A benefit of option 3 is that it can happen either through a properly authorised app endpoint, or even through the governance mechanism (propose, vote, accept), and is auditable in the ledger.

In some sense #1 is identical, assuming the HTML is updated through the same mechanism, but relies on browser/client-side checking. I do not have a good sense of whether that's widely implemented or not.

Reply:
I have seen mixed usage of option 1 and option 3.